### PR TITLE
OIDC: try to JWT decode userinfo response if JSON parsing failed

### PR DESCRIPTION
**Pull request by MatMaul**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by sandhose** [comment]
`userinfo` returning a JWT is well specced in OIDC Core: https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse

> If the UserInfo Response is signed and/or encrypted, then the Claims are returned in a JWT and the content-type MUST be application/jwt.

You should rely on the Content-Type response header instead of doing this fallback

File: synapse/handlers/oidc.py [hunk 1]
```diff
@@ -818,14 +818,38 @@ async def _fetch_userinfo(self, token: Token) -> UserInfo:
...
+                "id_token_signing_alg_values_supported", ["RS256"]
+            )
+            jwt = JsonWebToken(alg_values)
+            jwk_set = await self.load_jwks()
+            try:
+                decoded_resp = jwt.decode(body, key=jwk_set)
+            except ValueError:
+                logger.info("Reloading JWKS after decode error")
+                jwk_set = await self.load_jwks(force=True)  # try reloading the jwks
+                decoded_resp = jwt.decode(body, key=jwk_set)
```
Comment (reivilibre) [inline synapse/handlers/oidc.py:846]: can you explain what this reloading is about?

**Review by reivilibre — COMMENTED** [review]
looks reasonable otherwise, I'm just confused about this one piece

File: synapse/handlers/oidc.py (hunk 1 above)
Comment (MatMaul) [reply synapse/handlers/oidc.py:846]: If we can't decrypt the JWT payload, we try to reload the keys. I didn't invent it, it's in `_verify_jwt` :)

I tried to use `_verify_jwt` at the time and failed, but I am not sure why :thinking: 

Let me try again, because I should be able to just use it by passing `claims_cls=UserInfo`.

Comment (MatMaul) [reply synapse/handlers/oidc.py:846]: I tried again and I can't use `UserInfo` for `claims_cls`, claims have a specific format and expectation (cf `JWTClaims`) and `UserInfo` is not a claim.

Comment (reivilibre) [reply synapse/handlers/oidc.py:846]: I can't say I like this magic, but if you didn't invent it, then it's not really your problem :D

**Review by reivilibre — APPROVED** [review]
